### PR TITLE
[New transform] Add `KSpaceSpikeNoise` for MRI spike artifacts

### LINKED ISSUE #335: [New transform] Add `KSpaceSpikeNoise` for MRI spike artifacts
## Motivation

Isolated high-energy points in MRI k-space produce structured stripe/ringing artifacts in reconstructed images. Ordinary pixel-space salt-and-pepper noise or `RingingOvershoot` does not model the same acquisition failure. TorchIO exposes this family as `RandomSpike`.

Sources:

- [TorchIO augmentation transforms](https://docs.torchio.org/2.0/reference/transforms/spike/)
- [TorchIO paper](https://www.sciencedirect.com/science/article/pii/S0169260721003102)

## Proposed transform

Add `KSpaceSpikeNoise`, an image/volume-only transform that injects one or more sampled spikes into the Fourier representation and reconstructs the result.

## Acceptance criteria

- Sample spike count, spatial-frequency locations, and relative intensity from explicit ranges.
- Define whether conjugate-symmetric spikes are inserted to guarantee a real reconstruction, or how residual imaginary values are handled.
- FFT/IFFT operate only over spatial axes; batch and channel dimensions are excluded.
- Provide shared-versus-independent channel semantics and deterministic seeded replay/serialization.
- Preserve output shape and documented dtype/range for `float32` and `uint8`.
- Test zero strength, DC/near-DC handling, Nyquist and boundary locations, odd/even shapes, non-cubic 3D inputs, `D=1`, multiple spikes, and finite outputs.
- Include phantom/impulse reference tests that verify artifact orientation from known spike coordinates.
- Benchmark runtime and peak memory and reuse a single transform-domain representation when injecting multiple spikes.
- Documentation distinguishes k-space spikes from image-space impulse noise and convolution ringing.

## Out of scope

- General differentiable Fourier augmentation policies.
- Learning spike distributions from scanner data inside the transform.

## TorchIO API baseline (audited 2026-08-19)

TorchIO `Spike` exposes:

- `num_spikes`: scalar or sampled integer range; default `1`
- `intensity`: scalar or sampled range relative to the spectrum maximum; default `0.0`

It injects point impulses into the Fourier spectrum and reconstructs the image with `torch.fft`. The inherited transform policy includes `p`, `copy`, `per_instance`, `include`, and `exclude`.

AX currently has no public k-space spike transform. `SaltAndPepper`, impulse-like pixel noise, and ringing transforms do not reproduce this acquisition artifact.

### PR DESCRIPTION
**Pull request by DhruvGarg111**
Closes #335

## Summary

Adds `A.KSpaceSpikeNoise`, an image/volume-only transform that samples k-space spike artifacts (TorchIO `RandomSpike` family) by injecting sharp impulses into the Fourier spectrum and reconstructing the result.

- Samples spike count, relative intensity, and full-grid frequency locations from explicit ranges (`num_spikes_range`, `intensity_range`; locations uniform over `[0, N_i)` per spatial axis).
- Injects conjugate-symmetric real amplitudes so the half-spectrum stays Hermitian: the reconstruction is real by construction via `irfftn` — no imaginary residual. Self-conjugate bins (DC, Nyquist of even axes) are injected once; odd spatial sizes pass the explicit output shape to `irfftn`.
- FFT/IFFT operate over spatial axes only; batch and channel dimensions are excluded. One transform-domain representation is reused for all spikes: a single FFT, one `abs`/max pass, in-place injections, one IFFT.
- Shared or independent per-channel spike semantics (`per_channel`), one realization per invocation reused across a batch, and the volume processed as a single 3D transform.
- Seeded deterministic sampling and replay: realized scalar ranges recorded in applied configuration; realized spikes retained in-memory for exact replay; JSON round-trip reruns with resampled coordinates (RUNNABLE).
- uint8 processed as float32 in `[0, 1]` and converted back with rounding; outputs stay in `[0, 255]`, float32 clipped to `[0, 1]`.
- Fourier kernel routed through `torch.fft` (already a hard runtime import of the library), chosen over numpy/scipy by the complete-route benchmark matrix below.

## Coverage

- Functional closed-form reference tests verify artifact orientation from known spike coordinates: cosine stripes along rows/columns, DC constant shift `a/(H*W)`, Nyquist `cos(pi*x)*cos(pi*y)` with a single (not doubled) injection, halved-axis fold collision `(0, W-1)`, and a 3D plane wave along depth; diff-spectrum assertion isolates injected bins.
- Focused transform tests cover zero-intensity and zero-spike exact identities (uint8 + float32), odd/even and non-cubic shapes, `D=1` volumes, mismatched image/volume channel counts, shared vs per-channel semantics, batch/volume reuse, seeded reproducibility, replay reuse of realized spikes, serialization round-trip, applied-range recording, extreme intensity, finite outputs, and constructor validation.
- Contract base + `shared-spikes` extended cases registered; image/volume family, batch, and functional-3D benchmark matrices extended; benchmark coverage aliases added and `benchmark_coverage check` green.
- README, public docstring (distinguishing k-space spikes from `SaltAndPepper` image-space impulse noise and `RingingOvershoot` image-domain convolution ringing), and positional-parameter metadata updated.

One interpretation to flag: "spatial-frequency locations from explicit ranges" is implemented as uniform sampling over the full frequency grid (per-axis bounds are explicit and documented, matching TorchIO). A user-facing band-restriction parameter was intentionally not added.

## Benchmarks

Complete-route backend comparison (float32, 4 spikes, ms/call, same machine and venv, numpy<->torch bridge included in the torch column):

| size x channels | numpy | scipy | torch |
|---|---|---|---|
| 256x256x1 | 1.7 | 0.7 | 0.8 |
| 256x256x3 | 8.9 | 2.8 | 1.7 |
| 256x256x5 | 16.7 | 6.3 | 3.0 |
| 512x512x1 | 13.4 | 5.6 | 2.2 |
| 512x512x3 | 42.8 | 15.0 | 7.2 |
| 512x512x5 | 61.1 | 21.7 | 11.0 |
| 1024x1024x1 | 46.0 | 13.4 | 8.2 |
| 1024x1024x3 | 165.3 | 65.9 | 34.9 |
| 1024x1024x5 | 251.8 | 83.8 | 55.2 |

Torch wins 8 of 9 cells (256^2x1 ties at 0.8 vs scipy 0.7 ms), 1.5-2.5x over scipy and 3-4.5x over numpy on the MRI-relevant sizes. Rejected candidates recorded: numpy (single-threaded pocketfft) and scipy (threaded but slower than torch across the matrix). Public `Compose` route at 512x512x3: ~7.5 ms/call including dispatch.

Peak memory (tracemalloc, 1024x1024x3): the rfftn half-spectrum stays complex64 — ~36 MiB vs ~144 MiB for a full complex128 `fftn` (4x reduction). ASV matrices cover the direct kernel, image/volume family, and batch routes.

## Validation

- `tests/test_k_space_spike_noise.py`: 34 passed
- `tests/contracts`: 1856 passed, 6 skipped
- Target-cluster contracts: 1011 passed
- `ruff check` + `ruff format --check`, Pyrefly 0 errors, `check-ax-coding-guidance` clean, `tools.benchmark_coverage check` green, `git diff --check` clean
- Full fast quality gate: exit 0

## AI assistance

This PR was developed with AI assistance (Hermes Agent by Nous Research). Implementation, tests, and benchmarks were produced with automated tooling and verified locally with the validation commands above before opening; all findings described in this body were measured on the local environment.

## Summary by Sourcery

Introduce `KSpaceSpikeNoise` to simulate structured MRI spike artifacts through configurable Fourier-domain perturbations.

New Features:
- Add the public `KSpaceSpikeNoise` transform for generating MRI k-space spike artifacts on images and volumes, with configurable spike counts, intensities, and channel behavior.

Enhancements:
- Support deterministic sampling, batch and volume reuse, replay, serialization, dtype preservation, clipping, and real-valued Fourier reconstruction for the new transform.

Documentation:
- Document the new transform in the README and public API documentation, including how it differs from image-space impulse noise and image-domain ringing.

Tests:
- Add functional, contract, replay, serialization, edge-case, and benchmark coverage for k-space spike artifacts across image, batch, and 3D volume inputs.

### CODE REVIEW

**Comment by DhruvGarg111** [comment]
I have read and agree to the AlbumentationsX CLA Version 2.0 (July 14, 2026) as an individual.

File: albumentations/augmentations/pixel/noise.py [hunk 1]
```diff
@@ -1649,3 +1652,189 @@ def _sample_noise(
...
+        >>> import numpy as np
+        >>> import albumentations as A
+        >>> image = np.random.randint(0, 256, (128, 128, 3), dtype=np.uint8)
+        >>> transform = A.Compose(
+        ...     [A.KSpaceSpikeNoise(num_spikes_range=(2, 4), intensity_range=(0.1, 0.3), p=1.0)],
+        ...     seed=137,
+        ... )
+        >>> spiked = transform(image=image)["image"]
+
+        Volumes are supported with spikes that are invariant along depth:
```
[the remark this answers is not in the document]
Comment (DhruvGarg111) [reply albumentations/augmentations/pixel/noise.py]: Resolved: the Examples line was reworded to 'Volumes are supported with one 3D spike realization per invocation' (the Notes section already described the 3D transform). Folded into feat(transform).

File: benchmark/benchmarks/test_functional_kernels.py [hunk 2]
```diff
@@ -658,6 +664,8 @@ def setup(self, case_id: str) -> None:
         if name == "rician_noise":
             self.real_noise = rng.standard_normal(self.volume.shape, dtype=np.float32) * 0.1
             self.imaginary_noise = rng.standard_normal(self.volume.shape, dtype=np.float32) * 0.1
+        if name == "k_space_spike":
+            self.spikes = rng.integers(0, min(depth, height, width), size=(4, 3))
```
[the remark this answers is not in the document]
Comment (DhruvGarg111) [reply benchmark/benchmarks/test_functional_kernels.py]: Resolved: the benchmark fixture now samples per-axis bounds via np.array([depth, height, width]) instead of min(depth, height, width), so non-cubic volumes exercise the full coordinate range. Folded into the test/perf commit.

File: tools/benchmark_coverage.py [hunk 3]
```diff
@@ -392,6 +395,7 @@
     "PixelDropout": ("pixel_dropout",),
     "Posterize": ("posterize",),
     "RicianNoise": ("rician_noise",),
+    "KSpaceSpikeNoise": ("k_space_spike",),
     "StochasticConvolution": ("stochastic_convolve_shared", "stochastic_convolve_per_channel"),
```
[the remark this answers is not in the document]
Comment (DhruvGarg111) [reply tools/benchmark_coverage.py:413]: Resolved: KSpaceSpikeNoise is now registered in DIRECT_KERNEL_TRANSFORMS; tests/test_benchmark_coverage.py (21 passed) and the benchmark coverage check are green locally. This was the compatibility-job CI failure and is fixed on the pushed branch.

File: albumentations/augmentations/pixel/_functional_noise.py [hunk 4]
```diff
@@ -795,3 +798,69 @@ def rician_noise(
...
+    if not injections:
+        return img
+
+    ndim = spikes.shape[-1]
+    is_batch = img.ndim == ndim + 2
+    axes = tuple(range(1, ndim + 1)) if is_batch else tuple(range(ndim))
+    axis_sizes = tuple(img.shape[axis] for axis in axes)
+
+    tensor = torch.from_numpy(np.ascontiguousarray(img))
+    spectrum = torch.fft.rfftn(tensor, dim=axes)
```
[the remark this answers is not in the document]
Comment (DhruvGarg111) [reply albumentations/augmentations/pixel/_functional_noise.py]: Not triggered in the supported paths, verified empirically: tensor image (C,H,W), batch (C,L,H,W), and volume (C,D,H,W) inputs all route through the framework tensor bridge, which converts to channel-last NumPy before apply and restores the tensor layout afterward (validated with Compose calls on all three target types on this branch). The HWC/channel-last NumPy contract matches every other functional kernel in this module (e.g. rician_noise, add_noise, convolve) and the documented Compose contract; the kernel is not a public entry point for raw CHW tensors, and KSpaceSpikeNoise is not in PYTORCH_NATIVE_TENSOR_TRANSFORMS. No change needed.

**Comment by DhruvGarg111** [comment]
CI note: the Correctness gate failure is a runner-load artifact, not a code failure. The three cancelled jobs (Primary coverage, Compatibility ubuntu 3.11/3.12) carry a hard `timeout-minutes: 15` in .github/workflows/pr.yml and were timed out mid-run under the current runner load; every job that completed is SUCCESS, on both pushed heads (17/17 compatibility shards, contracts, ruff/mypy/pyrefly, ASV benchmark evidence, PyTorch tests, CLA). The previously failing benchmark-coverage test is fixed and verified locally (21/21). A maintainer-side rerun of the cancelled jobs should clear the gate.

**Comment by ternaus** [comment]
@codex

**Comment by ternaus** [comment]
@codex
